File: sdk/approute-sdk/python/approute/resources/accounts.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from approute.models import AccountListResponse, AccountTransactionPage
from approute.resources._base import BaseResource


class AccountsResource(BaseResource):

    def balances(self) -> AccountListResponse:
        """List account balances."""
        data = self._t.request("GET", "/accounts")
        return AccountListResponse.model_validate(data)

    def transactions(
        self,
        *,
        currency: str | None = None,
        category: list[str] | None = None,
        search: str | None = None,
        transaction_id: str | None = None,
        order_id: str | None = None,
        date_from: datetime | None = None,
        date_to: datetime | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> AccountTransactionPage:
        """List balance transactions with filtering."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if currency is not None:
            params["currency"] = currency
        if category is not None:
            params["category"] = category
        if search is not None:
            params["search"] = search
        if transaction_id is not None:
            params["transaction_id"] = transaction_id
        if order_id is not None:
            params["order_id"] = order_id
        if date_from is not None:
            params["date_from"] = date_from.isoformat()
        if date_to is not None:
            params["date_to"] = date_to.isoformat()
        data = self._t.request("GET", "/accounts/transactions", params=params)
        return AccountTransactionPage.model_validate(data)

File: sdk/approute-sdk/javascript/src/resources/accounts.ts
import { BaseResource } from "./base-resource.js";
import type {
  AccountListResponse,
  AccountTransactionPage,
} from "../models/index.js";

export interface AccountTransactionsOptions {
  currency?: string;
  category?: string[];
  search?: string;
  transactionId?: string;
  orderId?: string;
  dateFrom?: string;
  dateTo?: string;
  limit?: number;
  offset?: number;
}

/**
 * Resource for the /accounts endpoints (balances and transactions).
 */
export class AccountsResource extends BaseResource {
  /**
   * List account balances.
   */
  async balances(): Promise<AccountListResponse> {
    return this.transport.request<AccountListResponse>("GET", "/accounts");
  }

  /**
   * List balance transactions with filtering.
   */
  async transactions(
    options?: AccountTransactionsOptions,
  ): Promise<AccountTransactionPage> {
    const params: Record<string, unknown> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };

    if (options?.currency !== undefined) params.currency = options.currency;
    if (options?.category !== undefined) params.category = options.category;
    if (options?.search !== undefined) params.search = options.search;
    if (options?.transactionId !== undefined)
      params.transactionId = options.transactionId;
    if (options?.orderId !== undefined) params.orderId = options.orderId;
    if (options?.dateFrom !== undefined) params.dateFrom = options.dateFrom;
    if (options?.dateTo !== undefined) params.dateTo = options.dateTo;

    return this.transport.request<AccountTransactionPage>(
      "GET",
      "/accounts/transactions",
      { params },
    );
  }
}
